src/appendable_data.rs

```diff
@@ -0,0 +1,542 @@
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct PubPermissions {
+    permissions: BTreeMap<User, PubPermissionSet>,
+    /// The current index of the data when this permission change happened
+    data_index: u64,
+    /// The current index of the owners when this permission change happened
+    owner_entry_index: u64,
+}
+
+impl PubPermissions {
+    fn check_anyone_permissions(&self, action: Action) -> bool {
+        match self.permissions.get(&User::Anyone) {
+            None => false,
+            Some(perms) => perms.is_allowed(action).unwrap_or(false),
+        }
+    }
+}
+
+impl Permissions for PubPermissions {
+    fn is_action_allowed(&self, requester: PublicKey, action: Action) -> bool {
+        match self.permissions.get(&User::Key(requester)) {
+            Some(perms) => perms
+                .is_allowed(action)
+                .unwrap_or_else(|| self.check_anyone_permissions(action)),
+            None => self.check_anyone_permissions(action),
+        }
+    }
+
+    fn data_index(&self) -> u64 {
+        self.data_index
+    }
+
+    fn owner_entry_index(&self) -> u64 {
+        self.owner_entry_index
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct Owners {
+    owners: PublicKeySet,
+    /// The current index of the data when this ownership change happened
+    data_index: u64,
+    /// The current index of the permissions when this ownership change happened
+    permission_entry_index: u64,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+struct AppendOnly<P: Permissions> {
+    name: XorName,
+    tag: u64,
+    data: Vec<(Vec<u8>, Vec<u8>)>,
+    permissions: Vec<P>,
+    owners: Vec<Owners>,
+}
+
+/// Common methods for all `AppendOnlyData` flavours.
+pub trait AppendOnlyData<P> {
+    // /// Get a list of permissions for the provided user from the last entry in the permissions list.
+    // fn user_permissions(&self, user: &User) -> Result<&PubPermissionSet, ClientError>;
+
+    /// Return a value for the given key (if it is present).
+    fn get(&self, key: &[u8]) -> Option<&Vec<u8>>;
+
+    /// Get a list of keys and values with the given indexes.
+    fn at(&self, start: Index, end: Index) -> Option<&[(Vec<u8>, Vec<u8>)]>;
+
+    /// Return all entries.
+    fn entries(&self) -> &Vec<(Vec<u8>, Vec<u8>)>;
+
+    /// Return the name of this AppendOnlyData.
+    fn name(&self) -> XorName;
+
+    /// Return the type tag of this AppendOnlyData.
+    fn tag(&self) -> u64;
+
+    /// Return the last entry index.
+    fn entry_index(&self) -> u64;
+
+    /// Return the last owners index.
+    fn owners_index(&self) -> u64;
+
+    /// Return the last permissions index.
+    fn permissions_index(&self) -> u64;
+
+    /// Get a complete list of permissions from the entry in the permissions list at the specified index.
+    fn permissions(&self, start: Index, end: Index) -> Option<&[P]>;
```

Comment:
We can perhaps consider including `range` in this function and in the `owners()` function too